Interactive Render Motion Blur

danieldresser-ie opened the pull request:
Not quite 100% finished, but I think it's very reasonable to review in this state.

There are two tests failing:  a new test for updating the shutter override on a camera interactively, which is just one last update sychronize that I haven't done yet, and I have broken testNonInterpolableObjectSamples somehow, hopefully neither of those will be particularly tricky fixes.

In terms of structure, the main questions are around the signature of transformSamples and objectSamples.  Needing to pass around both the vector of hashes and vector of times results in some pretty messy behaviours:  transformSamples sets sampleTimes, unless sampleHashes is set, in which case it reads sampleTimes, though if it turns out that all the samples are identical even though they hashed differently then it also writes sampleTimes again.  See if you've got any ideas on how to make this tidier ( I don't really want to have to duplicate setting up sampleTimes though ... ).

There's also the question of needing to standardize between transformSamples and RenderController::SceneGraph::updateTransform on whether to take ScenePlug* or M44fPlug*.  I generally prefer to avoid passing in information to a function that isn't used, so I've currently gone with M44fPlug, but I don't care much if you want me to go the other way.

johnhaddon commented on src/GafferScene/RenderController.cpp:
```diff
@@ -316,6 +330,35 @@ class RenderController::SceneGraph
+				}
+			}
+
+			if( changedGlobals & TransformBlurGlobalComponent )
+			{
+				m_dirtyComponents |= TransformComponent;
+			}
```
I presume there is a reason why we need to do this even if we discovered above that `m_transformSegments` wasn't actually changing? It's not obvious to me though, so I think needs an explanatory comment.

danieldresser-ie replied:
I think this was left over from an earlier version and is no longer necessary.  Removed in 6a932f6

johnhaddon commented on src/GafferScene/RenderController.cpp:
```diff
@@ -316,6 +330,35 @@ class RenderController::SceneGraph
+				m_dirtyComponents |= TransformComponent;
+			}
+
+			if( changedGlobals & DeformationBlurGlobalComponent )
```
As above.

danieldresser-ie replied:
Removed in 6a932f6

johnhaddon commented on src/GafferScene/RenderController.cpp:
```diff
@@ -344,7 +387,7 @@ class RenderController::SceneGraph
 			if( ( m_dirtyComponents & TransformComponent ) || parentTransformChanged )
 			{
-				if( updateTransform( controller->m_scene->transformPlug(), parentTransformChanged ) )
+				if( updateTransform( controller->m_scene->transformPlug(), parentTransformChanged, controller->m_motionBlurOptions ) )
```
When I first saw this I wondered what the interaction between `m_transformSegments` and `controller->m_motionBlurOptions` was going to be, since we already used the latter in the computation of the former. Looks like we only use the `shutter` part of the options in `updateTransform()` though, so perhaps we should pass that alone?

Actually, since we already store `m_transformTimes`, can we just compute it earlier where we're currently updating `m_transformSegments`, and ditch `m_transformSegments` entirely? Seems like the dataflow is a bit more obvious that way potentially? Probably worth considering how this impacts the discussion around the interaction between hashes and samples as well.

danieldresser-ie replied:
I've gone with your direction of ditching m_transformSegments in c060dee.  It has quite a lot of knock-on effects - most quite positive.

The one slightly weird part:  we now don't modify m_transformTimes in transformSamples - we instead just ignore the times when a single sample is returned ( ie if samples are all identical ).  This keeps m_transformTimes around so we can use it for future evaluations when things might not all be identical.  But not modifying m_transformTimes means we also need a pointer m_transformTimesOutput, which tracks whether we are using the times from m_parent.

johnhaddon commented on src/GafferScene/RenderController.cpp:
```diff
@@ -354,7 +397,7 @@ class RenderController::SceneGraph
 			// Object
 
-			if( ( m_dirtyComponents & ObjectComponent ) && updateObject( controller->m_scene->objectPlug(), type, controller->m_renderer.get(), controller->m_globals.get(), controller->m_scene.get(), controller->m_lightLinks.get() ) )
+			if( ( m_dirtyComponents & ObjectComponent ) && updateObject( controller->m_scene->objectPlug(), type, controller->m_renderer.get(), controller->m_globals.get(), controller->m_scene.get(), controller->m_lightLinks.get(), controller->m_motionBlurOptions ) )
```
Same comment as above - I think we only need `motionBlurOptions.shutter` here.

danieldresser-ie replied:
Now irrevelevant after c060dee

johnhaddon commented on include/GafferScene/RenderController.h:
```diff
@@ -96,7 +96,26 @@ class GAFFERSCENE_API RenderController : public boost::signals::trackable
+			AllGlobalComponents = GlobalsGlobalComponent | SetsGlobalComponent | RenderSetsGlobalComponent | CameraOptionsGlobalComponent | TransformBlurGlobalComponent | DeformationBlurGlobalComponent
+		};
+
+		struct MotionBlurOptions
+		{
+			MotionBlurOptions() : transformBlur( false ), deformationBlur( false ), shutter( 0 )
+			{
+			}
+
+			MotionBlurOptions( const IECore::CompoundObject *globals, const ScenePlug *scene );
+
+			bool operator==( const MotionBlurOptions &other )
+			{
+				return transformBlur == other.transformBlur && deformationBlur == other.deformationBlur && shutter == other.shutter;
+			}
+			bool transformBlur;
+			bool deformationBlur;
+			Imath::V2f shutter;
```
I wonder if this is a place where the inline initialisers for data members are worth using? Maybe makes the definition a little more concise?

```
struct MotionBlurOptions
{
	MotionBlurOptions() = default;
	MotionBlurOptions( const IECore::CompoundObject *globals, const ScenePlug *scene );
	bool transformBlur = false;
	bool deformationBlur = false;
	Imath::V2f shutter = Imath::V2f( 0 );
};
```

danieldresser-ie replied:
Done in 6d00f94

johnhaddon reviewed (commented):
Thanks Daniel! Comments in line as usual, except I haven't looked in detail at the changes to `updateObject()` and `updateTransform()` as I suspect that would be a little premature given your comment about the signatures for `transformSamples/objectSamples/transformHashSamples`.

> In terms of structure, the main questions are around the signature of transformSamples and objectSamples. Needing to pass around both the vector of hashes and vector of times results in some pretty messy behaviours: transformSamples sets sampleTimes, unless sampleHashes is set, in which case it reads sampleTimes, though if it turns out that all the samples are identical even though they hashed differently then it also writes sampleTimes again. See if you've got any ideas on how to make this tidier ( I don't really want to have to duplicate setting up sampleTimes though ... ).

A collection of functions with tricky interdependencies surrounding shared data implies "that should be a class" to me. Being able to encapsulate some of that logic separately would also be pretty beneficial to readability in the SceneGraph code, since it's pretty dense already. Not sure if that helps?

```
TransformSampler
{
    void update( const M44fPlug *plug, const CompoundAttributes *fullAttributes, MotionBlurOptions &, ObjectInterface *object );
};
```

> There's also the question of needing to standardize between transformSamples and RenderController::SceneGraph::updateTransform on whether to take ScenePlug* or M44fPlug*. I generally prefer to avoid passing in information to a function that isn't used, so I've currently gone with M44fPlug, but I don't care much if you want me to go the other way.

I think that's a sound choice.

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+
+			self.assertEqual( len( samples ), len( sampleTimes ) )
+			for (i,j) in zip( samples, sampleTimes ):
+				self.assertAlmostEqual( i, j, places = 6 )
```
Not particularly surprising to me - I'm using some samples that are 0.2 apart, which isn't exactly representable.  Splitting the range [0.6, 1.4] into quarters I would definitely expect to introduce some sort of floating point error, and we're probably not being particularly careful about it.  Why wouldn't you expect things to be off by an epsilon at least?

johnhaddon replied:
I was thinking it wouldn't matter what the error was, because the values we're sampling are driven exactly by `context.getFrame()`, so should have the exact same value as was used for the shutter open/close. But you're right, the expected values are coming from Python, where they're computed differently _and_ represented as doubles.

danieldresser-ie commented on src/GafferScene/RenderController.cpp:
```diff
@@ -596,30 +658,65 @@ class RenderController::SceneGraph
+				{
+					m_fullTransform.reserve( samples.size() );
+
+					vector<M44f>::const_iterator s = samples.begin();
```
Fixed in 6a932f6

danieldresser-ie commented on src/GafferScene/RenderController.cpp:
```diff
@@ -1318,6 +1572,24 @@ void RenderController::updateInternal( const ProgressCallback &callback, const I
+
+			MotionBlurOptions newOptions( m_globals.get(), m_scene.get() );
+			if( !( newOptions == m_motionBlurOptions ) )
+			{
```
FIxed in 6d00f94.

I went perhaps even further then you were suggesting, and move all logic out of MotionBlurOptions.  I'm not sure why I thought it would make things cleaner, but I didn't like how it turned out.

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+				samples = [ i.radius() for i in capturedSphere.capturedSamples() ]
+				times = capturedSphere.capturedSampleTimes()
+				#for i, r in enumerate( sampleTimes ) :
+					#self.assertEqual( samples[i].radius(), r )
```
Fixed in 6a932f6

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+				samples = [ i.translation().x for i in capturedSphere.capturedTransforms() ]
+				times = capturedSphere.capturedTransformTimes()
+				#for i, r in enumerate( sampleTimes ) :
+					#self.assertEqual( samples[i].radius(), r )
```
Fixed in 6a932f6

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+		# We get a single sample out even if the transform hash is changing but the transform isn't
+		sphere["transform"]["translate"]["x"].setInput( dummyFrame["output"] )
+		controller.update()
+		assertMotionSamples( [ 3 ], False )
```
Renamed to expectedSamples in 6a932f6

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+		camera['renderSettingOverrides']['shutter']["value"].setValue( imath.V2f( -0.5, 0.5 ) )
+		options['options']['renderCamera']["enabled"].setValue( True )
+		options['options']['renderCamera']["value"].setValue( "/group/camera" )
+		# TODO - these tests currently failing, which I think represents a genuine failure to track camera updates
```
Fixed in 6d00f94 using CameraShutterGlobalComponent

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+		sphereAttributes['attributes']['transformBlur']["value"].setValue( False )
+		sphereAttributes['attributes']['deformationBlur']["value"].setValue( False )
+		controller.update()
+		#assertMotionSamples( [ 0.6, 0.8, 1.0, 1.2, 1.4 ], False )
```
Fixed in 6a932f6

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+		sphereAttributes['attributes']['transformBlurSegments']["enabled"].setValue( True )
+		sphereAttributes['attributes']['transformBlurSegments']["value"].setValue( 1 )
+		controller.update()
+		# Very counter-intuitively, this does nothing, because the sphere is not moving
```
Fair enough that based on a hierachy based mental model this would be the expected result - what I don't like it that it can switch back and forth depending on whether animated samples of the child happen to be identical within this shutter. 
 Imagine a helicopter blade where the parent is spinning, and the child is animating tilt.  It would be quite confusing when on frame where the tilt changes, you lose all your motion blur samples - whereas if the child segments always overrode, it wouldn't match a hiearchy based mental model, but at least it would be predictable and easy to explain.

johnhaddon replied:
That's not how the existing batch output works though is it? If a child has non-changing samples, we're keeping the samples from the parent, and concatenating the single child transform with each one.

danieldresser-ie replied:
I'm describing the same thing as how batch currently works.  If the child has non-changing samples, we keep the parent samples, if it has changing samples, we overwrite them.  So depending on whether an animated object happens to move on a frame, we may get completely different sampling.

johnhaddon replied:
Right, got you - I had completely misread what you were saying. Yep, that is definitely counterintuitive. Interesting that it hasn't come up in practice though. I think we're OK to keep current behaviour for this PR, and we're in a better position to change behaviour in future now RendererAlgo has been privatised.

danieldresser-ie commented on python/GafferSceneTest/RenderControllerTest.py:
```diff
@@ -527,5 +527,217 @@ def testNullObjects( self ) :
+		# Very counter-intuitively, this does nothing, because the sphere is not moving
+		assertMotionSamples( [ 0.6, 0.8, 1.0, 1.2, 1.4 ], False )
+
+		# But then if the sphere moves, the sample count does take affect
```
I was starting to lean towards the generating merged samples in the name of correctness, but then I found a much cleaner way of organising the function signatures for dealing with sampleTimes that wouldn't work with that, so hopefully we don't want to do that.

danieldresser-ie commented on src/GafferScene/RenderController.cpp:
```diff
@@ -316,6 +330,35 @@ class RenderController::SceneGraph
 					}
 				}
+
+				if( ( m_changedComponents & AttributesComponent ) || ( changedGlobals & ( TransformBlurGlobalComponent | DeformationBlurGlobalComponent ) ) )
```
The redundant work thing I have fixed.  Lifting out of the root I haven't done yet - it will be easy to do, but I just want to understand why - these settings aren't inherited ( the inheritance uses m_fullAttributes ), they're only used for setting transforms or objects at the current location, and I don't think it's valid to do that at the root?  Maybe we want to possibly allow for transforms at the root in the future?

johnhaddon replied:
I was thinking this was necessary in order to correctly inherit changes to blur settings from the global attributes. But I now see this inheritance is already taken care of by the `parentAttributesChanged` flag, which will cause us to do `m_changedComponents |= AttributesComponent` and enter the code via the path you've already got. I've tested this interactively too, and all is well.

danieldresser-ie commented on src/GafferScene/RenderController.cpp:
```diff
@@ -422,11 +473,14 @@ class RenderController::SceneGraph
 				( m_dirtyComponents & BoundComponent )
 			)
 			{
+				// TODO - what is this used for?  Does it need accurate motion blur?
```
Removed comment in 6a932f6

danieldresser-ie replied:
Added it back - I think it's in the right place now.

danieldresser-ie commented:
In addition to addressing your feedback, I did take a swing at making RendererAlgo private - which turned out to be enough of a rabbit hole that perhaps I should have deferred it, but now that I'm making a significant change to the transformSamples signature, it feels weird to have to announce an API change, and then afterwards announce that the API is removed.

While this was a bit of a rabbit hole, I think I'm moderately happy with the result - I think it makes sense to move the Python bindings for RendererAlgo into GafferSceneTest, since I can't think of any legitimate usage of them other than testing?  And then we can just use whatever signature is convenient for testing?

Is it OK to take a couple of things private that are used by Capsule?  Technically it's fine, they're in the same module.  Conceptually ... I think it's also fine?  We're not expecting downstream customers of Gaffer to implement something as closely tied to Gaffer's renderer system as Capsule, right?

johnhaddon commented on src/GafferScene/RenderController.cpp:
```diff
@@ -823,20 +972,55 @@ class RenderController::SceneGraph
 
 		IECore::MurmurHash m_transformHash;
-		Imath::M44f m_fullTransform;
+		std::vector<Imath::M44f> m_fullTransform;
```
Might be worth using `small_vector` here and for `m_deformationTimes`? Typically we're only expecting two floats...

danieldresser-ie replied:
Sounds kinda reasonable, but we're not saving anything if we have to convert to vector, right?  Are you proposing changing the interface to IECoreScenePreview::Renderer::ObjectInterface::transform?  Might be reasonable, but probably one for another time?

johnhaddon replied:
Good point - yep, we can leave for another time.

johnhaddon reviewed (changes requested):
Thanks Daniel - did a bit of interactive testing with this today and it's really nice to see motion blur updating interactively. As well as the repro for the bug I mailed you, I've made a few comments inline - hopefully it all makes sense...
Cheers...
John

danieldresser-ie commented on include/GafferScene/SceneAlgo.h:
```diff
@@ -267,6 +271,26 @@ GAFFERSCENE_API bool visible( const ScenePlug *scene, const ScenePlug::ScenePath
 /// for other object types we must return a synthetic bound.
 GAFFERSCENE_API Imath::Box3f bound( const IECore::Object *object );
 
+/// Render Adaptors
+/// ===============
+
+/// Function to return a SceneProcessor used to adapt the
+/// scene for rendering.
+typedef std::function<SceneProcessorPtr ()> RenderAdaptor;
+/// Registers an adaptor.
+GAFFERSCENE_API void registerRenderAdaptor( const std::string &name, RenderAdaptor adaptor );
+/// Removes a previously registered adaptor.
+GAFFERSCENE_API void deregisterRenderAdaptor( const std::string &name );
+/// Returns a SceneProcessor that will apply all the currently
+/// registered adaptors.
+GAFFERSCENE_API SceneProcessorPtr createRenderAdaptors();
+
+/// Apply Camera Globals
+/// ====================
+
+/// Applies the resolution, aspect ratio etc from the globals to the camera.
+GAFFERSCENE_API void applyCameraGlobals( IECoreScene::Camera *camera, const IECore::CompoundObject *globals, const ScenePlug *scene );
```
Done ( squashed in place )

johnhaddon replied:
We've released `0.60.0.0b1` now, so the changes need to be rebased and put in a new `0.60.0.0bx` section at the top. I can take care of this if you would prefer - I know the changes are a pain.

danieldresser-ie commented on include/GafferScene/Private/RendererAlgo.h:
```diff
@@ -58,6 +58,9 @@ namespace GafferScene
 
 IE_CORE_FORWARDDECLARE( SceneProcessor )
 
+namespace Private
+{
+
```
Done ( squashed in place )

danieldresser-ie commented on src/GafferSceneTestModule/GafferSceneTestModule.cpp:
```diff
@@ -44,19 +44,57 @@
+void testOutputCameras( const GafferScene::ScenePlug &scene, const IECore::CompoundObject &globals, const GafferScene::ScenePlug &renderSetsScene, IECoreScenePreview::Renderer &renderer )
+{
+	GafferScene::Private::RendererAlgo::RenderSets renderSets( &renderSetsScene );
+	IECorePython::ScopedGILRelease gilRelease;
```
Switched to old bindings.  I've done the binding in src/GafferSceneModule/RenderBinding.cpp, since that was currently where we create GafferScene.Private, and that seemed like a relevant enough name.

danieldresser-ie commented:
OK, I think I've addressed everything, other than the render glitch you provided me with a repro for.

It certainly seemed like I just must have added a bug where I was accessing uninitialized memory, but after adding a whole pile of checks, I couldn't find anything.  What instead started to find was a lot of evidence that this may be an Arnold issue:

* it doesn't happen with a sphere by itself, but the trigger isn't being in a group, it is having another object existing in the scene
* changing the size of the plane causes a refresh that fixes the issue - but I'm pretty sure that I don't do any edits to the sphere when the plane changes ... this implies that I have already sent the data correctly to Arnold, since it is somehow able to recover without more data
* the artifact looks like flat axis-aligned cutouts being taken out of the correct motion blur shape - even when the moving shape is a sphere - there shouldn't be any data we could pass to Arnold that would cause this
* the artifact occurs not just when changing segments, but also when changing frames and respecifying motion, even with 1 segment blur ( which seems like a simpler code path on our end )

Spent a bunch of time looking for a bug in my stuff, but starting to think this looks pretty consistent with Arnold's BVH not getting rebuilt.  I should probably try and create a simpler repro independent of Gaffer to test, but wanted to get the rest of this PR taken care of first.

johnhaddon commented:
Thanks! The explanation regarding the render glitch definitely sounds plausible. I tried the render in 3Delight to see if that could corroborate, but it is broken too, albeit in a different way. Appleseed, on the other hand, gives the results we would expect. Seems worth trying to get a repro together to prove beyond doubt that it's Arnold though?

danieldresser-ie commented:
I've now reproduced two artifacts with the Arnold Python API - one when I change the number of segments, the other when I change the shutter.  I'm waiting to hear from SolidAngle whether this is an issue on their end, or if I'm we're somehow using the API wrong ( but the API calls in my simple example look extremely reasonable ).

johnhaddon commented:
Thanks - just pushed a rebase to sort Changes.md out, and will merge once CI passes.
